coding_style/README.adoc

```diff
@@ -0,0 +1,217 @@
+  The same can go for many other modules that can be given an entire list of items all at once.
+* Use meta modules when possible.
++
+[%collapsible]
+====
+Rationale:: This will allow our playbooks to run on the widest selection of operating systems possible without having to modify any more tasks than is necessary.
+Examples::
+* Instead of using the `upstart` and `systemd` modules, use the `service`
+module when at all possible.
+* Similarly for package management, use `package` instead of `yum` or `dnf` or
+similar.
+====
++
+* Avoid the use of `lineinfile` wherever that might be feasible.
++
+[%collapsible]
+====
+Rationale:: Slight miscalculations in how it is used can lead to a loss of idempotence.
+Modifying config files with it can cause the Ansible code to become arcane and difficult to read, especially for someone not familiar with the file in question.
+Try editing files directly using other built-in modules (e.g. `ini_file`, `blockinfile`, `xml`), or reading and parsing.
+If you are modifying more than a tiny number of lines or in a manner more than trivially complex, try leveraging the `template` module, instead.
+This will allow the entire structure of the file to be seen by later users and maintainers.
+The use of `lineinfile` should include a comment with justification.
+====
++
+* Limit use of the `copy` module to copying remote files and to uploading binary blobs.
+  For all other file pushes, use the `template` module. Even if there is nothing in the file that is being templated at the current moment, having the file handled by the `template` module now makes adding that functionality much simpler than if the file is initially handled by the `copy` and then needs to be moved before it can be edited.
+# TODO: I am not convinced about the following rule:
+* When using the `template` module, refrain from appending `.j2` to the file name. This alters the syntax highlighting in most editors and will obscure the benefits of highlighting for the particular file type or filename.
+  Anything under the `templates` directory of a role is assumed to be treated as a Jinja 2 template, so adding the `.j2` extension is redundant information that is not helpful.
+* Keep filenames and templates as close to the name on the destination system as possible.
++
+[%collapsible]
+====
+Rationale:: This will help with both editor highlighting as well as identifying source and destination versions of the file at a glance.
+Avoid duplicating the remote full path in the role directory, however, as that creates unnecessary depth in the file tree for the role.
+Grouping sets of similar files into a subdirectory of `templates` is allowable, but avoid unnecessary depth to the hierarchy.
+====
++
+* Use `{{ role_path }}/subdir/` as the filename prefix when including files if the name has a variable in it.
```

Comment:
That's specific to roles and belongs in this section.

Reply:
moved.